platform/linux-generic/odp_packet_io.c

```diff
@@ -2851,11 +2944,37 @@ uint64_t odp_pktin_wait_time(uint64_t nsec)
 	return (nsec / (1000)) + 1;
 }
 
+static void check_tx_compl_ev(const odp_packet_hdr_t *hdr, int pkt_idx, tx_compl_info_t *info,
+			      uint16_t *num)
+{
+	if (odp_unlikely(hdr->p.flags.tx_compl)) {
+		info[*num].user_ptr = hdr->user_ptr;
+		info[*num].queue = hdr->dst_queue;
+		info[*num].idx = pkt_idx;
+		(*num)++;
+	}
+}
+
+static void send_tx_compl_events(tx_compl_info_t *info, uint16_t num, odp_buffer_t bufs[],
+				 int num_sent)
+{
+	for (int i = 0; i < num; i++) {
+		if (info[i].idx < num_sent)
+			_odp_pktio_send_tx_compl_event(bufs[i], info[i].user_ptr, info[i].queue);
+		else
+			odp_buffer_free(bufs[i]);
```

Comment:
When we hit else there won't be anymore compl events to send (all following idx will be larger). We can call buffer_free_multi and break.

odp_buffer_free_multi(&bufs[i], num - i);
break;

Reply:
Done in v6.